src/algorithms/calorimetry/CalorimeterClusterRecoCoG.cc

```diff
@@ -161,3 +161,159 @@ void CalorimeterClusterRecoCoG::AlgorithmProcess() {
+  float totalE = 0.;
+  float maxE   = 0.;
+  // Used to optionally constrain the cluster eta to those of the contributing hits
+  float minHitEta = std::numeric_limits<float>::max();
+  float maxHitEta = std::numeric_limits<float>::min();
+  auto time       = pcl->getHits()[0].getTime();
+  auto timeError  = pcl->getHits()[0].getTimeError();
+  for (unsigned i = 0; i < pcl->getHits().size(); ++i) {
+    const auto& hit   = pcl->getHits()[i];
+    const auto weight = pcl->getWeights()[i];
+    m_log->debug("hit energy = {} hit weight: {}", hit.getEnergy(), weight);
+    auto energy = hit.getEnergy() * weight;
+    totalE += energy;
+    if (energy > maxE) {
+    }
+    const float eta = edm4eic::eta(hit.getPosition());
+    if (eta < minHitEta) {
+      minHitEta = eta;
+    }
+    if (eta > maxHitEta) {
+      maxHitEta = eta;
+    }
+  }
+  cl.setEnergy(totalE / m_sampFrac);
+  cl.setEnergyError(0.);
+  cl.setTime(time);
+  cl.setTimeError(timeError);
+
+  // center of gravity with logarithmic weighting
+  float tw = 0.;
+  auto v   = cl.getPosition();
+  for (unsigned i = 0; i < pcl->getHits().size(); ++i) {
+    const auto& hit   = pcl->getHits()[i];
+    const auto weight = pcl->getWeights()[i];
+    //      _DBG_<<" -- weight = " << weight << "  E=" << hit.getEnergy() << " totalE=" <<totalE << " log(E/totalE)=" << std::log(hit.getEnergy()/totalE) << std::endl;
+    float w           = weightFunc(hit.getEnergy() * weight, totalE, m_logWeightBase, 0);
+    tw += w;
+    v = v + (hit.getPosition() * w);
+  }
+  if (tw == 0.) {
+    m_log->warn("zero total weights encountered, you may want to adjust your weighting parameter.");
+  }
+  cl.setPosition(v / tw);
+  cl.setPositionError({}); // @TODO: Covariance matrix
+
+  // Optionally constrain the cluster to the hit eta values
+  if (m_enableEtaBounds) {
+    const bool overflow  = (edm4eic::eta(cl.getPosition()) > maxHitEta);
+    const bool underflow = (edm4eic::eta(cl.getPosition()) < minHitEta);
+    if (overflow || underflow) {
+      const double newEta   = overflow ? maxHitEta : minHitEta;
+      const double newTheta = edm4eic::etaToAngle(newEta);
+      const double newR     = edm4eic::magnitude(cl.getPosition());
+      const double newPhi   = edm4eic::angleAzimuthal(cl.getPosition());
+      cl.setPosition(edm4eic::sphericalToVector(newR, newTheta, newPhi));
+      m_log->debug("Bound cluster position to contributing hits due to {}", (overflow ? "overflow" : "underflow"));
+    }
+  }
+
+  // Additional convenience variables
+
+  // best estimate on the cluster direction is the cluster position
+  // for simple 2D CoG clustering
+  cl.setIntrinsicTheta(edm4eic::anglePolar(cl.getPosition()));
+  cl.setIntrinsicPhi(edm4eic::angleAzimuthal(cl.getPosition()));
+  // TODO errors
+
+  //_______________________________________
+  // Calculate cluster profile:
+  //    radius,
+  //   	dispersion (energy weighted radius),
+  //    eta-phi cluster widths (2D)
```

Comment:
I would like to point out that eta is not given by `edm4eic::anglePolar`. A possible resolution could be:
```suggestion
  //    theta-phi cluster widths (2D)
```

Reply:
Yes, simple misprint in the comments.